### PR TITLE
Add bt coex jam

### PR DESCRIPTION
**Pull request by Tschet1**
This adds device-tree and KConfig support for the 1wire bluetooth coexistence protocol. 
This updates the documentation for the bluetooth coexistence protocol.
This adds a sample for the 1wire coexistence protocol, where the user can control the granting/denial of radio activity through a simple uart console.

This relies on https://github.com/nrfconnect/sdk-nrf/pull/6117

### CODE REVIEW

File: subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/*
```
Comment (jori-nordic) [inline subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c:1]: This does not belong in the DTS commit, since it's really initializing the whole module. The PR's commit history is a bit hard to understand/review in general. 

I'd prefer if you put all the changes for enabling 1-wire coex into 1 commit (you could even put the sample in), and renaming of the kconfig/sources into separate commits.
DTS could be put in its own commit but then please don't include initialization code with it (ie. only have the yaml files).

File: subsys/mpsl/cx/CMakeLists.txt [hunk 2]
```diff
@@ -8,5 +8,5 @@ zephyr_library()
 
 zephyr_library_sources_ifdef(CONFIG_MPSL_CX mpsl_cx.c)
 zephyr_library_sources_ifdef(CONFIG_MPSL_CX_THREAD thread/mpsl_cx_thread.c)
-zephyr_library_sources_ifdef(CONFIG_MPSL_CX_BT_3WIRE bluetooth/mpsl_cx_bluetooth.c)
+zephyr_library_sources_ifdef(CONFIG_MPSL_CX_BT_3WIRE bluetooth/mpsl_cx_3w_bluetooth.c)
```
Comment (jori-nordic) [inline subsys/mpsl/cx/CMakeLists.txt:11]: this should be a fixup of another commit or belong in a "Rename 3-wire coex" or something commit where you put all the kconfig/filename renaming

File: include/mpsl/mpsl_cx_config_bluetooth.h [hunk 3]
```diff
@@ -18,7 +18,7 @@
 
 #include <stdint.h>
 
-/** @brief Configures the Bluetooth Radio Coexistence interface.
+/** @brief Configures the Bluetooth Radio Three Wire Coexistence interface.
```
Comment (rugeGerritsen) [inline include/mpsl/mpsl_cx_config_bluetooth.h:21]: This change seems unintentional

File: subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+#include <drivers/gpio.h>
+#include <devicetree.h>
+#include <nrfx.h>
+#include <nrfx_ppi.h>
+#include <nrfx_gpiote.h>
+#include <mpsl_coex.h>
+
+#define COEX_TIMER NRF_TIMER1
+
+#if DT_HAS_COMPAT_STATUS_OKAY(sdc_radio_coex_one_wire)
```
Comment (jori-nordic) [inline subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c:25]: we use a different way now, because upstream asked for it. 
You have to use a pointer in the radio node to the coex node, and resolve that pointer instead.
This is how do for 3-wire coex:
```c
#if DT_NODE_HAS_STATUS(DT_PHANDLE(DT_NODELABEL(radio), coex), okay)
#define COEX_NODE DT_PHANDLE(DT_NODELABEL(radio), coex)
#else
#define COEX_NODE DT_INVALID_NODE
#error No enabled coex nodes registered in DTS.
#endif

#if !(DT_NODE_HAS_COMPAT(COEX_NODE, sdc_radio_coex_three_wire))
#error Selected coex node is not compatible with sdc-radio-coex-three-wire.
#endif
```

File: include/mpsl/mpsl_cx_config_bluetooth.h [hunk 5]
```diff
@@ -30,6 +30,18 @@
...
+ *
+ * This function sets device interface parameters for the Coexistence module.
+ * The module is used to control PTA interface through the given pins and resources.
+ *
+ * @retval   0             Coexistence interface successfully configured.
+ * @retval   -NRF_EPERM    Coexistence interface is not available.
+ *
+ */
+
+int32_t mpsl_cx_bt_interface_1wire_config_set(void);
```
Comment (rugeGerritsen) [inline include/mpsl/mpsl_cx_config_bluetooth.h:43]: I have a slight preference to move `mpsl_cx_bt_interface_1wire_config_set` above the 3-wire interface as it is simpler.

File: subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		return -ENOMEM;
+	}
+	if (nrfx_ppi_channel_alloc(&ppi_channel) != NRFX_SUCCESS) {
+		return -ENOMEM;
+	}
+	gpiote_cfg->gpio_pin = DT_GPIO_PIN(COEX_NODE, grant_gpios);
+	gpiote_cfg->active_high = MPSL_COEX_BT_GPIO_POLARITY_GET(grant_gpios);
+	gpiote_cfg->ppi_ch_id = ppi_channel;
+
+    /* update concurrency mode */
```
Comment (jori-nordic) [inline subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c:66]: identation

File: subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+	/* Allocate GPIOTE and PPI channels for the GRANT line */
+	gpiote_cfg = &coex_if->grant_cfg;
+	if (nrfx_gpiote_channel_alloc(&gpiote_cfg->gpiote_ch_id) != NRFX_SUCCESS) {
+		return -ENOMEM;
+	}
+	if (nrfx_ppi_channel_alloc(&ppi_channel) != NRFX_SUCCESS) {
+		return -ENOMEM;
+	}
+	gpiote_cfg->gpio_pin = DT_GPIO_PIN(COEX_NODE, grant_gpios);
```
Comment (jori-nordic) [inline subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c:62]: Martí made me aware of a pitfall of using this, it won't work for pins on the Port 1.
Use this instead
```c
NRF_DT_GPIOS_TO_PSEL(COEX_NODE, grant_gpios);
```

File: samples/bluetooth/radio_coex_1wire/README.rst [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+This sample demonstrates how to use the 1-Wire coexistence feature described in :ref:`Bluetooth® external radio coexistence <nrfxlib:bluetooth_coex>`.
+
+Overview
+********
+
+The application sets up an advertiser and configures the 1wire coexistence feature.
+The user can enter ``g`` to grant or ``d`` to deny Bluetooth radio activity.
+
+* When Bluetooth radio activity is granted, packets will be transmitted.
+* When Bluetooth radio activity is, no packets will be transmitted.
```
Comment (jori-nordic) [inline samples/bluetooth/radio_coex_1wire/README.rst:19]: is denied

File: samples/bluetooth/radio_coex_1wire/README.rst [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+.. include:: /includes/build_and_run.txt
+
+Testing
+=======
+
+1. Build and program the development kit.
+#. |connect_terminal|
+#. Watch the number of packets transmitted.
+#. Enter ``d`` to deny packet transmission.
+#. Observe that there are no packets transmitted, anymore.
```
Comment (jori-nordic) [inline samples/bluetooth/radio_coex_1wire/README.rst:54]: no more packets transmitted. Or `Observe that packet transmissions have stopped`. And `Observe that packet transmission has re-started` or `started again`

File: samples/bluetooth/radio_coex_1wire/prj.conf [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+
+CONFIG_NCS_SAMPLES_DEFAULTS=y
+CONFIG_ASSERT=y
+
+CONFIG_BT=y
+CONFIG_BT_PERIPHERAL=y
+CONFIG_BT_CENTRAL=y
```
Comment (jori-nordic) [inline samples/bluetooth/radio_coex_1wire/prj.conf:12]: you are not doing any central activity

File: samples/bluetooth/radio_coex_1wire/src/main.c [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+#include <hal/nrf_radio.h>
+#include <hal/nrf_timer.h>
+#include <helpers/nrfx_gppi.h>
+#include <nrfx_gpiote.h>
+#include <nrfx_timer.h>
+#include <nrfx_ppi.h>
+#include <hal/nrf_gpio.h>
+
+#define STACKSIZE CONFIG_MAIN_STACK_SIZE
```
Comment (jori-nordic) [inline samples/bluetooth/radio_coex_1wire/src/main.c:22]: is this define really needed ?

File: samples/bluetooth/radio_coex_1wire/src/main.c [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+#define THREAD_PRIORITY K_LOWEST_APPLICATION_THREAD_PRIO
```
Comment (jori-nordic) [inline samples/bluetooth/radio_coex_1wire/src/main.c:23]: is this really needed ?

File: samples/bluetooth/radio_coex_1wire/src/main.c [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+{
+	if (console_init()) {
+		__ASSERT(false, "Failed to initialise console");
+	}
+
+	printk("-----------------------------------------------------\n");
+	printk("This sample illustrates the 1Wire coex interface\n");
+	printk("The number of radio request every second is printed\n");
+	printk("\n");
+	printk("Press a key to change state:\n");
```
Comment (jori-nordic) [inline samples/bluetooth/radio_coex_1wire/src/main.c:51]: I'd put `change PTA behavior` or `grant behavior`.

File: samples/bluetooth/radio_coex_1wire/src/main.c [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+
+static void print_welcome_message(void)
+{
+	if (console_init()) {
+		__ASSERT(false, "Failed to initialise console");
+	}
+
+	printk("-----------------------------------------------------\n");
+	printk("This sample illustrates the 1Wire coex interface\n");
+	printk("The number of radio request every second is printed\n");
```
Comment (jori-nordic) [inline samples/bluetooth/radio_coex_1wire/src/main.c:49]: The number of radio events is printed every second. (or.. sent radio packets ..)

**Review by jori-nordic — CHANGES_REQUESTED** [review]

**Review by rugeGerritsen — CHANGES_REQUESTED** [review]

File: samples/bluetooth/radio_coex_1wire/README.rst [hunk 15]
```diff
@@ -0,0 +1,71 @@
+.. _bluetooth_radio_coex_1wire_sample:
+
+Bluetooth: External radio coexistence using one-wire interface
+################################################################
+
+.. contents::
+   :local:
+   :depth: 2
+
+This sample demonstrates how to use the 1-Wire coexistence feature described in :ref:`Bluetooth® external radio coexistence <nrfxlib:bluetooth_coex>`.
```
Comment (wiba-nordic) [inline samples/bluetooth/radio_coex_1wire/README.rst:10]: Wire doesn't need to be capitalized here. 

Also, the title of the page uses "one-wire", and here is "1-wire". It's better to stick to using one or the other.

File: samples/bluetooth/radio_coex_1wire/README.rst [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+Overview
+********
+
+The application sets up an advertiser and configures the 1wire coexistence feature.
```
Comment (wiba-nordic) [inline samples/bluetooth/radio_coex_1wire/README.rst:15]:
```suggestion
The application sets up an advertiser and configures the 1-wire coexistence feature.
```
Missing hyphen.

File: samples/bluetooth/radio_coex_1wire/README.rst (hunk 8 above)
Comment (wiba-nordic) [inline samples/bluetooth/radio_coex_1wire/README.rst:19]:
```suggestion
* When Bluetooth radio activity is denied, no packets will be transmitted.
```
Missing word

File: samples/bluetooth/radio_coex_1wire/README.rst (hunk 9 above)
Comment (wiba-nordic) [inline samples/bluetooth/radio_coex_1wire/README.rst:54]:
```suggestion
#. Observe that there are no packets transmitted anymore.
```

File: samples/bluetooth/radio_coex_1wire/README.rst [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+Testing
+=======
+
+1. Build and program the development kit.
+#. |connect_terminal|
+#. Watch the number of packets transmitted.
+#. Enter ``d`` to deny packet transmission.
+#. Observe that there are no packets transmitted, anymore.
+#. Enter ``g`` to grant packet transmission.
+#. Observe that packets are transmitted, again.
```
Comment (wiba-nordic) [inline samples/bluetooth/radio_coex_1wire/README.rst:56]:
```suggestion
#. Observe that packets are transmitted again.
```

**Review by wiba-nordic — CHANGES_REQUESTED** [review]

File: subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+	((GPIO_ACTIVE_LOW & DT_GPIO_FLAGS(COEX_NODE, dt_property)) ? false : true)
+
+static volatile bool coex_enabled;
+
+static void coex_enable_callback(void)
+{
+	coex_enabled = true;
+}
+
+int mpsl_cx_bt_interface_1wire_config_set(void)
```
[the remark this answers is not in the document]
Comment (Tschet1) [reply subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c:46]: this is called in mpsl_cx.c if 1wire coex is enabled.

File: include/mpsl/mpsl_cx_config_bluetooth.h (hunk 3 above)
Comment (Tschet1) [reply include/mpsl/mpsl_cx_config_bluetooth.h:21]: I think it is intentional. Do you see any issues with it?

Comment (rugeGerritsen) [reply include/mpsl/mpsl_cx_config_bluetooth.h:21]: Sorry, I see this is now docs for the three wire implementation

File: samples/bluetooth/radio_coex_1wire/src/main.c (hunk 12 above)
Comment (Tschet1) [reply samples/bluetooth/radio_coex_1wire/src/main.c:23]: not really, I guess.

File: subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+#else
+#define COEX_NODE DT_INVALID_NODE
+#error No "sdc-radio-coex-one-wire"-compatible (and enabled) nodes found
+#endif
+
+#if !IS_ENABLED(CONFIG_SOC_SERIES_NRF52X)
+#error Bluetooth coex is only supported on the nRF52 series
+#endif
+
+#define MPSL_COEX_BT_GPIO_POLARITY_GET(dt_property)                                                \
```
[the remark this answers is not in the document]
Comment (Tschet1) [reply subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c:36]: we could, but I would prefer to have it in this file, then it is clearer to look at the 3W part, in my opinion. I think you would then have all the configuration in one place instead of having half of it in the header an the other half in the c file.

[the remark this answers is not in the document]
Comment (Tschet1) [reply subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c:36]: Then we would need to define in the C file the COEX_NODE first before including the header. I think that's less clean.

[the remark this answers is not in the document]
Comment (jori-nordic) [reply subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c:36]: alright, no strong opinion :)

File: subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c (hunk 1 above)
Comment (Tschet1) [reply subsys/mpsl/cx/bluetooth/mpsl_cx_1w_bluetooth.c:1]: updated

File: include/mpsl/mpsl_cx_config_bluetooth.h (hunk 5 above)
Comment (Tschet1) [reply include/mpsl/mpsl_cx_config_bluetooth.h:43]: will update also in other places to have it aligned

File: samples/bluetooth/radio_coex_1wire/src/main.c [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+static void console_print_thread(void)
+{
+	uint32_t *task_capture =
+		(uint32_t *)nrf_timer_task_address_get(M_COUNTER, nrf_timer_capture_task_get(0));
+	uint32_t *task_clear =
+		(uint32_t *)nrf_timer_task_address_get(M_COUNTER, NRF_TIMER_TASK_CLEAR);
+	while (1) {
+		*task_capture = 1;
```
Comment (KAGA164) [inline samples/bluetooth/radio_coex_1wire/src/main.c:89]: Please use `nrf_timer_task_trigger(M_COUNTER, nrf_timer_capture_task_get(0));`

File: samples/bluetooth/radio_coex_1wire/src/main.c [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+
+		printk("Number of radio events last second: %d\n", nrf_timer_cc_get(M_COUNTER, 0));
+
+		*task_clear = 1;
```
Comment (KAGA164) [inline samples/bluetooth/radio_coex_1wire/src/main.c:93]: Please use Please use `nrf_timer_task_trigger(M_COUNTER, NRF_TIMER_TASK_CLEAR);`

File: samples/bluetooth/radio_coex_1wire/src/main.c (hunk 20 above)
Comment (KAGA164) [reply samples/bluetooth/radio_coex_1wire/src/main.c:89]: and remove lines 84-87

File: samples/bluetooth/radio_coex_1wire/src/main.c [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+#define APP_GRANT_GPIO_PIN NRF_DT_GPIOS_TO_PSEL(ZEPHYR_USER_NODE, coex_pta_grant_gpios)
+#else
+#error "Unsupported board: see README and check /zephyr,user node"
+#define APP_GRANT_GPIO_PIN 0
+#endif
+
+#define APP_GRANT_ACTIVE_LOW (GPIO_ACTIVE_LOW & DT_GPIO_FLAGS(COEX_NODE, grant_gpios) ? true : false)
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
```
Comment (KAGA164) [inline samples/bluetooth/radio_coex_1wire/src/main.c:43]: Don't you want to use device name in an advertising data ?

File: samples/bluetooth/radio_coex_1wire/src/main.c [hunk 23]
```diff
@@ -0,0 +1,161 @@
...
+static void console_print_thread(void)
+{
+	uint32_t *task_capture =
+		(uint32_t *)nrf_timer_task_address_get(M_COUNTER, nrf_timer_capture_task_get(0));
+	uint32_t *task_clear =
+		(uint32_t *)nrf_timer_task_address_get(M_COUNTER, NRF_TIMER_TASK_CLEAR);
+	while (1) {
+		*task_capture = 1;
+
+		printk("Number of radio events last second: %d\n", nrf_timer_cc_get(M_COUNTER, 0));
```
Comment (KAGA164) [inline samples/bluetooth/radio_coex_1wire/src/main.c:91]:
```suggestion
		printk("Number of radio events from last second: %d\n", nrf_timer_cc_get(M_COUNTER, 0));
```

File: samples/bluetooth/radio_coex_1wire/src/main.c (hunk 20 above)
Comment (KAGA164) [reply samples/bluetooth/radio_coex_1wire/src/main.c:89]: Also you can use a define for timer capture channel

**Review by KAGA164 — CHANGES_REQUESTED** [review]

**Review by wiba-nordic — APPROVED** [review]

File: subsys/mpsl/cx/CMakeLists.txt [hunk 24]
```diff
@@ -8,4 +8,4 @@ zephyr_library()
...
```
Comment (jori-nordic) [inline subsys/mpsl/cx/CMakeLists.txt:12]: nit: missing a word in commit message

File: dts/bindings/radio_coex/sdc-radio-coex-one-wire.yaml [hunk 25]
```diff
@@ -0,0 +1,18 @@
...
+compatible: "sdc-radio-coex-one-wire"
+
+include: generic-radio-coex-one-wire.yaml
+
+properties:
+    concurrency-mode:
+        type: int
+        required: true
+        description: |
+            Concurrency mode with the external modem supported by 1-wire coexistence.
```
[the remark this answers is not in the document]
Comment (jori-nordic) [reply dts/bindings/radio_coex/sdc-radio-coex-one-wire.yaml:18]: if we know it's always going to be either 1 or 0, we can use a binary property instead (like I did for rx level in 3-wire coex)

File: dts/bindings/radio_coex/generic-radio-coex-one-wire.yaml [hunk 26]
```diff
@@ -0,0 +1,17 @@
+# Copyright (c) 2021 Nordic Semiconductor ASA
```
Comment (jori-nordic) [inline dts/bindings/radio_coex/generic-radio-coex-one-wire.yaml:1]: commit msg nit: We're not only adding the configuration for 1-wire coex, but it seems like we're adding the whole feature in this commit: dts bindings, Kconfig options and init code. Could you also squash/fixup the documentation commit into this one ?

**Review by jori-nordic — APPROVED** [review]
A few nits regarding the commit history, but changes look good !

**Review by KAGA164 — APPROVED** [review]

**Review by rugeGerritsen — APPROVED** [review]
